lib/synapse/service_watcher/zookeeper.rb

```diff
@@ -134,6 +134,28 @@ def create(path)
       @zk.create(path, ignore: :node_exists)
     end
 
+    # get with retry to reduce failure when network connectivity is flaky or ZK is overloaded
+    def get(path, opts = {})
+      retry_limit = opts[:retry_limit] || 3
+      opts.delete(:retry_limit)
+      retry_interval = opts[:retry_interval] || 5
+      opts.delete(:retry_interval)
+      retry_count = 0
+      begin
+        if retry_count > retry_limit
+          raise RuntimeError, "synapse: get ZK path failed after retry"
+        end
+        retry_count += 1
+        node = @zk.get(path, opts)
+        raise IOError, "synapse: get ZK path return nil" if node.nil?
+        return node
+      rescue ZK::Exceptions::OperationTimeOut, IOError => e
```

Comment:
so we only `rescue` this two types of exception? or simply retry on all possible zk error? However, we should only retry on `zk` errors.

Reply:
these two are obviously retriable error, I was able to reproduce by cutting off the network. According to ZK API documentation [here](https://www.rubydoc.info/gems/zk/ZK/Client/Base#get-instance_method), it could also throw ZK::Exceptions::NoNode, but this means node does not exist any more, and we should instead do some cleanup. So here we just throw the error, and the call already has cleanup logics.

Reply:
the common practice is to differentiate retriable vs non-retriable errors, handle them separately.